### PR TITLE
Fix quadratic complexity in Reconciler.Reconcile().

### PR DESCRIPTION
**Pull request by Tener**
[`Reconciler.Reconcile()`](https://github.com/gravitational/teleport/blob/2588725623b9e824135637413b914418b5ed314c/lib/services/reconciler.go#L98) is ["accidentally quadratic"](https://news.ycombinator.com/item?id=9192127).
For large number of resources N, the running time of this function blows up:
```
    reconciler_test.go:299: [N=10000] Run time 710.941583ms, time for element 71.094µs
    reconciler_test.go:299: [N=20000] Run time 1.99310675s, time for element 99.655µs
    reconciler_test.go:299: [N=30000] Run time 4.042424417s, time for element 134.747µs
    reconciler_test.go:299: [N=40000] Run time 8.069562916s, time for element 201.739µs
    reconciler_test.go:299: [N=50000] Run time 13.143397s, time for element 262.867µs
    reconciler_test.go:299: [N=100000] Run time 1m13.327041833s, time for element 733.27µs 
```

This complexity arises because `Reconciler.Reconcile()` calls both `processNewResource` and `processRegisteredResource` N times, and each of those calls `currentResources.Find()` which also has N complexity. Combined we arrive at NxN=N^2.

After simple fix (switching to map lookup instead of linear list scan):
```
    reconciler_test.go:299: [N=10000] Run time 296.669375ms, time for element 29.666µs
    reconciler_test.go:299: [N=20000] Run time 575.412375ms, time for element 28.77µs
    reconciler_test.go:299: [N=30000] Run time 840.50825ms, time for element 28.016µs
    reconciler_test.go:299: [N=40000] Run time 1.104382709s, time for element 27.609µs
    reconciler_test.go:299: [N=50000] Run time 1.369955208s, time for element 27.399µs
    reconciler_test.go:299: [N=100000] Run time 2.700222125s, time for element 27.002µs
```

For increased efficiency and to make it clear reconciler operates on a list of _unique_ resources, I've changed the respective signatures to the new type `types.ResourcesWithLabelsMap`.

Additionally, two (unused) functions called `Find()` were removed due to the danger of accidentally repeating the problem in question. `collection.ToMap()["foo"]` should be used instead.

### CODE REVIEW

File: lib/srv/db/watcher.go [hunk 1]
```diff
@@ -134,11 +134,8 @@ func (s *Server) startCloudWatcher(ctx context.Context) error {
 }
 
 // getResources returns proxied databases as resources.
-func (s *Server) getResources() (resources types.ResourcesWithLabels) {
-	for _, database := range s.getProxiedDatabases() {
-		resources = append(resources, database)
-	}
-	return resources
+func (s *Server) getResources() (resources types.ResourcesWithLabelsMap) {
```
Comment (codingllama) [inline lib/srv/db/watcher.go:137]:
```suggestion
func (s *Server) getResources() types.ResourcesWithLabelsMap {
```

**Review by codingllama — APPROVED** [review]
Nice.

**Review by espadolini — APPROVED** [review]
what @codingllama said

File: api/types/resource.go [hunk 2]
```diff
@@ -90,14 +90,21 @@ type ResourceWithLabels interface {
 // ResourcesWithLabels is a list of labeled resources.
 type ResourcesWithLabels []ResourceWithLabels
```
Comment (zmb3) [inline api/types/resource.go:91]: Do we still need this type?

Comment (Tener) [reply api/types/resource.go:91]: Yes, we do: it is rather popular in the codebase. I could work to replace it, but not everywhere `ResourcesWithLabelsMap` makes sense.

**Pull request merged**